app/models/service_template.rb

```diff
@@ -403,6 +399,14 @@ def provision_request(user, options = nil, request_options = {})
     result[:request]
   end
 
+  def picture=(value)
+    if value
+      super unless value.kind_of?(Hash)
+
+      super(create_picture(value))
```

Comment:
How does this instance method get to the class method?

Reply:
As such, are there specs for this?

Reply:
it's the activerecord ```singular_association``` that's building it so we don't actually need the class method anyway. good try though.

Reply:
This's tested! It's in this PR and also at https://github.com/ManageIQ/manageiq/pull/18674/files#diff-7338735460ffd6eebe447a430abdec8eR1051. Beyond that I'm testing Rails since the method is here: https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/builder/singular_association.rb#L32

Reply:
Looks like it's calling the dynamic method for the singular association... My only concern is if self hasn't been persisted yet.  I don't know if `create_picture` would require it be saved first.  I don't know.  You might have to call Picture.new or something like that and pass that to super.  🤷‍♂

Reply:
```
irb(main):007:0* st = ServiceTemplate.create!(:name => "blah")
   (0.2ms)  BEGIN
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE ("users"."id" BETWEEN $1 AND $2) AND "users"."userid" IS NULL LIMIT $3  [["id", 0], ["id", 999999999999], ["LIMIT", 1]]
  User Inst Including Associations (0.0ms - 0rows)
  SQL (6.2ms)  INSERT INTO "service_templates" ("name", "guid", "created_at", "updated_at", "miq_group_id", "tenant_id", "internal", "service_type", "generic_subtype") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING "id"  [["name", "blah"], ["guid", "cb823fc0-8b3f-40a3-b195-ee71866d7539"], ["created_at", "2019-04-29 19:26:38.584945"], ["updated_at", "2019-04-29 19:26:38.584945"], ["miq_group_id", 1], ["tenant_id", 1], ["internal", "f"], ["service_type", "atomic"], ["generic_subtype", nil]]
   (0.4ms)  COMMIT
=> #<ServiceTemplate id: 44, name: "blah", description: nil, guid: "cb823fc0-8b3f-40a3-b195-ee71866d7539", type: nil, service_template_id: nil, options: {}, created_at: "2019-04-29 19:26:38", updated_at: "2019-04-29 19:26:38", display: nil, evm_owner_id: nil, miq_group_id: 1, service_type: "atomic", prov_type: nil, provision_cost: nil, service_template_catalog_id: nil, long_description: nil, tenant_id: 1, generic_subtype: nil, deleted_on: nil, internal: false, zone_id: nil>

irb(main):009:0> st.create_picture(:content => "iVBORw0KGgoAAAAN", :extension => "jpg")
   (0.2ms)  BEGIN
  SQL (3.6ms)  INSERT INTO "pictures" ("resource_id", "resource_type", "content", "extension", "md5") VALUES ($1, $2, $3, $4, $5) RETURNING "id"  [["resource_id", 44], ["resource_type", "ServiceTemplate"], ["content", "<16 bytes of binary data>"], ["extension", "jpg"], ["md5", "4617a55a374175d8d7479071f41db6c7"]]
   (0.4ms)  COMMIT
  Picture Load (1.0ms)  SELECT  "pictures".* FROM "pictures" WHERE "pictures"."resource_id" = $1 AND "pictures"."resource_type" = $2 LIMIT $3  [["resource_id", 44], ["resource_type", "ServiceTemplate"], ["LIMIT", 1]]
  Picture Inst Including Associations (0.1ms - 1rows)
=> #<Picture id: 15, resource_id: 44, resource_type: "ServiceTemplate", content: "iVBORw0KGgoAAAAN", extension: "jpg", md5: "4617a55a374175d8d7479071f41db6c7">
```

Kewl!

👍

Reply:
For what it is worth, this fails if the picture exists first:

```ruby
picture = Picture.create(:content => "foo", :extension => "jpg")
ServiceTemplate.create(:name => "foo", :picture => picture)
```

Currently used like this in specs:

https://github.com/ManageIQ/manageiq-api/blob/f12cf95d/spec/requests/service_templates_spec.rb#L16-L22